### PR TITLE
 Support IEEE754 floating point comparisons

### PR DESCRIPTION
**Pull request by PavelPunegov**
1. Update Compiler 

2. Support IEEE754
- Add an intrinsic that performs check according to the standard
- Add floating point compare methods from LLVM
- Support new IR builtins, symbols and descriptors
- Rewrite compareTo floating point checks to Kotlin
- Rewrite equals to be the same as in JVM

3. Unignore several tests

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt [hunk 1]
```diff
@@ -74,9 +69,14 @@ internal fun <T : CallableMemberDescriptor> T.resolveFakeOverride(): T {
 
 private val intrinsicAnnotation = FqName("konan.internal.Intrinsic")
 
-// TODO: check it is external?
 internal val CallableDescriptor.isIntrinsic: Boolean
-    get() = this.annotations.findAnnotation(intrinsicAnnotation) != null
+    get() = when {
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt]: Unneeded.

Comment (PavelPunegov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt]: At first I added the check for isExternal for FunctionDescriptor, but one of Igor's commits changed this extension to CallableDescriptor, which is not applicable for the isExternal check.
I'll remove this change

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 2]
```diff
@@ -2332,22 +2338,44 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
...
-        when (descriptor) {
-            ib.eqeqeq     -> return functionGenerationContext.icmpEq(args[0], args[1])
-            ib.gt0        -> return functionGenerationContext.icmpGt(args[0], kImmZero)
-            ib.gteq0      -> return functionGenerationContext.icmpGe(args[0], kImmZero)
-            ib.lt0        -> return functionGenerationContext.icmpLt(args[0], kImmZero)
-            ib.lteq0      -> return functionGenerationContext.icmpLe(args[0], kImmZero)
-            ib.booleanNot -> return functionGenerationContext.icmpNe(args[0], kTrue)
-            else -> {
-                TODO(descriptor.name.toString())
+        val funGen = functionGenerationContext
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Nope, use functionGenerationContext, maybe with apply/whatever.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 3]
```diff
@@ -2332,22 +2338,44 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
...
+            }
+            descriptor.isComparisonDescriptor(ib.lessOrEqualFunByOperandType) -> {
+                if (args[0].type.isFloatingPoint()) funGen.fcmpLe(args[0], args[1])
+                else funGen.icmpLe(args[0], args[1])
+            }
+            else -> TODO(descriptor.name.toString())
         }
     }
 
+    private fun LLVMTypeRef.isFloatingPoint(): Boolean {
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Belongs somewhere in ValueTypes.kt

**Pull request merged**